### DISCUSSION TITLE
What's the best approach to testing & mocking a Nuxt plugin?

### DISCUSSION THREAD

**Discussion by transreport-marley** [nuxt/nuxt#26426] [Questions]
Hey all 👋,

-----

_A little back story ✍️_

As of `nuxt@3.10.1` I noticed a few warnings in console that highlight that I have been using `$fetch` incorrectly. I had been using a custom composable that contained an `ofetch` instance which had various interceptors to handle `401` errors, as well as attaching the required `Authorization` header and so on. It then got a little bit more complicated as I started wrapping it in `useAsyncData` (even when not required) and then an over-engineer request queue 🤦 

To address the errors I took inspiration from a recommended approach in the Nuxt Docs, which was [Custom useFetch in Nuxt](https://notes.atinux.com/nuxt-custom-fetch). It is a similar approach to what I had, but stripped things back to basics.

 I only took it up to the point of the custom `$api` as that's all that was needed. In the event I need `useAsyncData` I will use it on a case by case basis.

_End of back story ✊_

-----

In practice the `$api` works perfectly in my pages/components. However, when it comes to writing the tests and attempting to mock the `$api` and the response that should be returned - I am at a loss.

The usage inside of the component requires the use of `useNuxtApp()`, so it looks something like this:
```ts
const { $api } from useNuxtApp();

try {
  const res = await $api('/foo');
  ...
} catch (error) {
  ...
}
```

I am using `nuxt`, `@nuxt/test-utils`, `@vue/test-utils` and `vitest` for the test setup.

Any advice would be much appreciated. Thanks! 🙇

**Comment by gpataciertocom** [comment]
I'm facing same issue. I don't know how to mock a plugin.
If the plugin is defined in a nuxt module, it becomes auto-imported, but, how could I mock it?
Thanks!

**Comment by RyanClementsHax** [comment]
FYI: just found out there is an open issue discussing this https://github.com/nuxt/test-utils/issues/544
